### DISCUSSION TITLE
WASM port

### DISCUSSION THREAD

**Discussion by ggerganov** [ggml-org/whisper.cpp#96] [Ideas]
We can easily build `whisper.cpp` as a WASM library using [Emscripten](https://emscripten.org):

```java
mkdir build-em
cd build-em
emcmake cmake ..
make
```

It looks like a big subset of SIMD intrinsics are already supported, so the performance might not be really bad:

https://emscripten.org/docs/porting/simd.html

So let's try running `whisper.cpp` directly in the browser!

- The model file could either be fetched on load, or the user can drag and drop it in the browser window
- We need a simple page that records a short audio at 16 kHz sampling rate and passes it to the WASM module for transcription. Probably something similar to this [ggwave example](https://github.com/ggerganov/ggwave/blob/master/examples/ggwave-js/index-tmpl.html#L102-L185) can be used

**Comment by richardburleigh** [comment]
Great to see you're looking at WASM!

I actually attempted it a few days ago but failed due to the medium model being too big to load (I know, obvious in hindsight). My use case is translation so the tiny model isn't applicable, but this is a really exciting development anyway.

**Comment by ggerganov** [comment]
The WASM example is now fully functional:

[examples/whisper.wasm](https://github.com/ggerganov/whisper.cpp/tree/master/examples/whisper.wasm)

The performance is better than I expected thanks to the WASM SIMD intrinsics.
It seems that the only browser that is not supported at the moment is Safari.

Would love to hear some feedback if someone gives it a try!
